### ISSUE TITLE
Operator runs out of ConfigMap space for collectors with ~> 450 PodMonitoring CRs

### ISSUE THREAD

**Issue by optix2000** [GoogleCloudPlatform/prometheus-engine#549]
We're running into an issue where we have a lot of PodMonitoring CR's and the collector is failing to scrape new namespaces.
After some debugging we traced it down to the gmp-operator failing to update the Configmap as the config has gotten too big for it.

Our use-case here is we want to have different scrape intervals between different namespaces as they have different levels of importance, so ClusterPodMonitoring doesn't work for us.

```
% kubectl get podmonitorings.monitoring.googleapis.com -A | wc -l
453
```

Operator logs are full of:
```
{"level":"error","ts":1692307167.247495,"logger":"controller.collector-config","msg":"Reconciler error","reconciler group":"monitoring.googleapis.com","reconciler kind":"OperatorConfig","name":"config","namespace":"gmp-public","error":"ensure collector config: update Prometheus config: ConfigMap \"collector\" is invalid: []: Too long: must have at most 1048576 bytes","errorVerbose":"ConfigMap \"collector\" is invalid: []: Too long: must have at most 1048576 bytes\nupdate Prometheus config\ngithub.com/GoogleCloudPlatform/prometheus-engine/pkg/operator.(*collectionReconciler).ensureCollectorConfig\n\t/app/pkg/operator/collection.go:330\ngithub.com/GoogleCloudPlatform/prometheus-engine/pkg/operator.(*collectionReconciler).Reconcile\n\t/app/pkg/operator/collection.go:178\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598\nensure collector config\ngithub.com/GoogleCloudPlatform/prometheus-engine/pkg/operator.(*collectionReconciler).Reconcile\n\t/app/pkg/operator/collection.go:179\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227"}
```

**Comment by macxamin** [comment]
Hello, we are working on supporting larger ConfigMap.

In the meantime, can you try simplifying your ConfigMap, possibly by using regex to simplify your `relabel_configs` and other sections where regex can be used.

**Comment by optix2000** [comment]
Our CR's are as simple as they get and we do not control the Configmap as it's managed by the gmp-operator.
```
apiVersion: monitoring.googleapis.com/v1
kind: PodMonitoring
metadata:
  name: app
  namespace: namespace
spec:
  endpoints:
  - interval: 120s
    port: 80
  selector:
    matchLabels:
      app: app
```

**Comment by JuddRogers** [comment]
There is an internal design for a fix out for review. ETA for fix is TBD.

**Comment by shinebayar-g** [comment]
We're facing the same issue.. Is there any known workaround?

**Comment by steakfest** [comment]
This same issue exists for the `rules-generated` config map, but the compression flag doesn't help there.   We ran into that issue today when deploying a collection of alert rules.

**Comment by shinebayar-g** [comment]
> This same issue exists for the `rules-generated` config map, but the compression flag doesn't help there. We ran into that issue today when deploying a collection of alert rules.

Oh no, I thought this PR was supposed to fix that configmap.

**Comment by dnck** [comment]
> Oh no, I thought this PR was supposed to fix that configmap.

After we added the compression: gzip feature to our operator config, our generated rules remained uncompressed. We tried to deploy new rules after the addition of the feature flag to no avail. If this PR was supposed to address it, either we're doing something funky over here, or perhaps this PR only had impact on the collection target rules.

**Comment by pintohutch** [comment]
Ah - thanks for raising here everyone. Indeed #571 (cc @TheSpiritXIII) only added compression for the collector configs. We hadn't anticipated needing it for the generated rules config, but it turns out that assumption was wrong.

We'll put up a fix to this!

**Comment by bwplotka** [comment]
Rules are not as easy as it seems: https://github.com/GoogleCloudPlatform/prometheus-engine/pull/763

TL;DR config-reloader has to have another feature - ability to uncompress additional watched files.

**Comment by dnck** [comment]
> Rules are not as easy as it seems: https://github.com/GoogleCloudPlatform/prometheus-engine/pull/763

Nice catch though! Glad ye didn't roll it to main. Thanks for the efforts on this one.
